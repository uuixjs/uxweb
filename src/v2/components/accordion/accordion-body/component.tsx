import { Background, BorderRadiusProps, Display, Layout } from "../../layout";
import { FC, ReactNode } from "react";

import { getAriaProps } from "lib";

export interface AccordionBodyProps {
  /**
   * @example Lorem ipsum dolor sit amet, consectetur adipiscing elit
   */
  children?: ReactNode;
  /**
   * @example true
   */
  isOpen?: boolean;
  borderRadius?: BorderRadiusProps;
}

/**
 * AccordionBody is a content region whose visibility should be
 * controlled by an AccordionHeader
 */
export const AccordionBody: FC<AccordionBodyProps> = (props) => (
  <Layout
    padding={2}
    background={Background.Base}
    elevation={1}
    display={props.isOpen ? Display.Block : Display.Hide}
    borderRadius={props.borderRadius}
    {...getAriaProps(props)}
  >
    {props.children}
  </Layout>
);

AccordionBody.displayName = "AccordionBody";
