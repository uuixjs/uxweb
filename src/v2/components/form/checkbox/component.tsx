import { ForwardRefRenderFunction, MutableRefObject, forwardRef } from "react";
import { INPUT_SIZES, InputControlProps, InputSize } from "../form";
import { Layout, Position } from "../../layout";
import { ScCheckBoxInputBase, ScCheckBoxLabelBase } from "../common";
import {
  cn,
  css,
  getAriaProps,
  getDataProps,
  newUUIDv4,
  omitDataProps,
  staticTokenRule,
  styled,
  themeTokenRule,
} from "lib/ui-utils";

import { withOverlayContext } from "../../overlay-region/context";

const ScCheckBoxLabel = styled(ScCheckBoxLabelBase)`
  &::before {
    border-radius: ${staticTokenRule("border-radius-small")};
  }

  ${/* sc-selector */ ScCheckBoxInputBase}:indeterminate + && {
    &::before {
      border-style: solid;
      border-width: ${staticTokenRule("border-width-input")};
      border-color: ${themeTokenRule("color-border-input-checkbox-checked")};
      background-color: ${themeTokenRule(
        "color-background-input-checkbox-checked-background",
      )};
    }

    &::after {
      background-color: ${themeTokenRule(
        "color-background-input-checkbox-checked",
      )};
      display: block;
      position: absolute;
      top: 50%;
      left: 0.4rem;
      width: 0.8rem;
      height: 0.2rem;
      transform: translateY(-50%);
      content: "";
    }

    /** Overlay style overrides */
    ${({ overlay }) =>
      overlay &&
      css`
        &::before {
          border-color: ${themeTokenRule(
            "color-border-input-checkbox-checked-overlay",
          )};
          background-color: ${themeTokenRule(
            "color-background-input-checkbox-checked-background-overlay",
          )};
        }

        &::after {
          background-color: ${themeTokenRule(
            "color-background-input-checkbox-checked-overlay",
          )};
        }
      `}
  }
`;

export interface CheckBoxProps extends InputControlProps {
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refDelegate` */
  refDelegate?: InputControlProps["refDelegate"];
  label: string | null;
  /**
   * Passes an `aria-label` to the component to provide context to assistive
   * devices. All standard HTML aria attributes are now supported on Core UI
   * components â€” see https://www.w3.org/TR/html-aria/."
   */
  "aria-label"?: string;
  size?: InputSize.Default | InputSize.Large | "default" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;;
  overlay?: boolean;
  indeterminate?: boolean;
}

export const CheckBoxComponent: ForwardRefRenderFunction<
  HTMLInputElement,
  CheckBoxProps
> = (
  {
    error,
    id,
    indeterminate,
    label,
    overlay,
    refDelegate,
    size = InputSize.Default,
    ...inputProps
  },
  ref,
) => {
  const generatedId = newUUIDv4();
  const classes: ClassValue = {
    "tw-checkbox": true,
    "tw-checkbox--error": error,
    "tw-checkbox--overlay": overlay,
  };
  const checkboxRef = (refValue: HTMLInputElement) => {
    if (refValue) {
      refValue.indeterminate = !!indeterminate;
    }

    if (typeof ref === "function") {
      ref(refValue);
    } else if (ref) {
      (ref as MutableRefObject<HTMLInputElement>).current = refValue;
    } else if (typeof refDelegate === "function") {
      refDelegate(refValue);
    } else if (refDelegate) {
      (refDelegate as MutableRefObject<HTMLInputElement>).current = refValue;
    }
  };

  return (
    <Layout
      className={cn(classes)}
      fontSize={INPUT_SIZES[size]}
      position={Position.Relative}
      {...getDataProps(inputProps)}
    >
      <ScCheckBoxInputBase
        {...omitDataProps(inputProps)}
        {...getAriaProps(inputProps)}
        className="tw-checkbox__input"
        type="checkbox"
        data-a-target="tw-checkbox"
        id={id ? id : generatedId}
        ref={checkboxRef}
        aria-label={label === undefined ? inputProps["aria-label"] : undefined}
      />
      <ScCheckBoxLabel
        className="tw-checkbox__label"
        htmlFor={id ? id : generatedId}
        overlay={overlay}
        error={error}
      >
        {label && <Layout margin={{ left: 1 }}>{label}</Layout>}
      </ScCheckBoxLabel>
    </Layout>
  );
};

CheckBoxComponent.displayName = "CheckBox";
const ComponentWithRef = forwardRef(CheckBoxComponent);
const ComponentWithOverlayContext = withOverlayContext<
  CheckBoxProps,
  HTMLInputElement
>(ComponentWithRef);
export { ComponentWithOverlayContext as CheckBox };
