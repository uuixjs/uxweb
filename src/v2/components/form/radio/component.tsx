import { ForwardRefRenderFunction, forwardRef } from "react";
import { INPUT_SIZES, InputControlProps, InputSize } from "../form";
import { Layout, Position } from "../../layout";
import { ScCheckBoxInputBase, ScCheckBoxLabelBase } from "../common";
import {
  cn,
  getAriaProps,
  getDataProps,
  newUUIDv4,
  omitDataProps,
  styled,
} from "lib/ui-utils";

import { withOverlayContext } from "../../overlay-region/context";

const ScRadioLabel = styled(ScCheckBoxLabelBase)`
  &::before,
  &::after {
    border-radius: 50%;
  }
`;

export interface RadioProps extends InputControlProps {
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refDelegate` */
  refDelegate?: InputControlProps["refDelegate"];
  label: string | null;
  /**
   * Passes an `aria-label` to the component to provide context to assistive
   * devices. All standard HTML aria attributes are now supported on Core UI
   * components â€” see https://www.w3.org/TR/html-aria/."
   */
  "aria-label"?: string;
  size?: InputSize.Default | InputSize.Large | "default" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;;
  overlay?: boolean;
}

export const RadioComponent: ForwardRefRenderFunction<
  HTMLInputElement,
  RadioProps
> = (
  {
    error,
    id,
    label,
    overlay,
    size = InputSize.Default,
    refDelegate,
    ...inputProps
  },
  ref,
) => {
  const generatedId = `tw-${newUUIDv4()}`;
  const classes: ClassValue = {
    "tw-radio": true,
    "tw-radio--error": error,
    "tw-radio--overlay": overlay,
  };

  return (
    <Layout
      className={cn(classes)}
      fontSize={INPUT_SIZES[size]}
      position={Position.Relative}
      {...getDataProps(inputProps)}
    >
      <ScCheckBoxInputBase
        {...omitDataProps(inputProps)}
        {...getAriaProps(inputProps)}
        ref={ref || refDelegate}
        type="radio"
        className="tw-radio__input"
        data-a-target="tw-radio"
        id={id ? id : generatedId}
        aria-label={label === undefined ? inputProps["aria-label"] : undefined}
      />
      <ScRadioLabel
        className="tw-radio__label"
        htmlFor={id ? id : generatedId}
        overlay={overlay}
        error={error}
      >
        {label && <Layout margin={{ left: 1 }}>{label}</Layout>}
      </ScRadioLabel>
    </Layout>
  );
};

RadioComponent.displayName = "Radio";
const ComponentWithRef = forwardRef(RadioComponent);
const ComponentWithOverlayContext = withOverlayContext<
  RadioProps,
  HTMLInputElement
>(ComponentWithRef);
export { ComponentWithOverlayContext as Radio };
