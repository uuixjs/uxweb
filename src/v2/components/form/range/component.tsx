import { Display, Layout, Position, ZIndex } from "../../layout";
import {
  FormEventHandler,
  ForwardRefRenderFunction,
  Ref,
  forwardRef,
} from "react";
import {
  css,
  focusVisible,
  getAriaProps,
  getDataProps,
  staticTokenRule,
  styled,
  themeTokenRule,
} from "lib";

import { StaticTokenMap } from "lib/ui-tokens";
import { withOverlayContext } from "../../overlay-region/context";

export interface RangePropsBase {
  disabled?: boolean;
  error?: boolean;
  id?: string;
  name?: string;
  onBlur?: FormEventHandler<HTMLInputElement>;
  onChange?: FormEventHandler<HTMLInputElement>;
  onFocus?: FormEventHandler<HTMLInputElement>;
  max?: number;
  min?: number;
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refDelegate` */
  refDelegate?: Ref<HTMLInputElement>;
  step?: number;
  tabIndex?: number;
  overlay?: boolean;
}

export interface RangePropsUnfilled extends RangePropsBase {
  defaultValue?: string;
  value?: string;
}

export interface RangePropsFilled extends RangePropsBase {
  fill: boolean;
  max: number;
  min: number;
  value: string;
}

export type RangeProps = RangePropsUnfilled | RangePropsFilled;

function isRangeFilled(props: RangeProps): props is RangePropsFilled {
  const filled = props as RangePropsFilled;
  return (
    filled.fill !== undefined &&
    filled.max !== undefined &&
    filled.min !== undefined &&
    filled.value !== undefined
  );
}

const trackHeight = "0.2rem";
const thumbHeight = "1.6rem";
const thumbWidth = "1.6rem";
const thumbMargin = "-0.7rem"; // "$track-height / 2 - $thumb-height / 2";

interface ScRangeProps {
  $overlay: boolean;
  $error: boolean;
  $filled: boolean;
}

const rangeSliderCss = css<ScRangeProps>`
  width: 100%;
  height: ${trackHeight};
  cursor: pointer;

  border-radius: ${staticTokenRule("border-radius-large")};

  background-color: ${({ $filled, $overlay }) =>
    $filled
      ? undefined
      : $overlay
      ? themeTokenRule("color-background-range-overlay")
      : themeTokenRule("color-background-range")};
`;

const rangeThumbCss = css<ScRangeProps>`
  border-radius: ${staticTokenRule("border-radius-rounded")};
  box-shadow: ${themeTokenRule("shadow-elevation-1")};
  width: ${thumbWidth};
  height: ${thumbHeight};
  margin-top: ${thumbMargin};
  border: ${staticTokenRule("border-width-default")} solid
    ${({ $overlay, $error }) =>
      $error && $overlay
        ? themeTokenRule("color-border-overlay-error")
        : $error
        ? themeTokenRule("color-border-input-error")
        : $overlay
        ? themeTokenRule("color-background-range-overlay-fill")
        : themeTokenRule("color-border-range-handle")};
  background-color: ${({ $overlay, $error }) =>
    $error && $overlay
      ? themeTokenRule("color-border-overlay-error")
      : $error
      ? themeTokenRule("color-border-input-error")
      : $overlay
      ? themeTokenRule("color-background-range-overlay-fill")
      : themeTokenRule("color-background-range-fill")};
  appearance: none;
`;

const rangeThumbFocusCss = css`
  border: ${staticTokenRule("border-width-input")} solid
    ${themeTokenRule("color-border-range-handle-focus")};
`;
const ScRangeInput = styled.input<ScRangeProps>`
  width: 100%;
  height: ${thumbHeight};
  vertical-align: middle;
  appearance: none;

  /** Intended to be applied to the full "clickable" area that can be dragged by the mouse */
  cursor: pointer;

  &,
  &:hover,
  &:focus,
  &:active {
    border: none;
    background: transparent;
    box-shadow: none;
  }

  &:disabled {
    opacity: 0.5;
    pointer-events: none;
  }

  /**
   * Slider
   */
  &::-moz-range-track {
    ${rangeSliderCss}
  }

  &::-ms-track {
    ${rangeSliderCss}

    /** MS Edge fails horribly when we try to use css vars on range psuedo elements. Avoid using any CSS vars on these!  */
    border-radius: ${StaticTokenMap["border-radius-large"]};
    background: ${StaticTokenMap["color-opac-b-4"]};
  }

  &::-webkit-slider-runnable-track {
    ${rangeSliderCss}
  }

  &::-ms-fill-lower {
    background-color: ${themeTokenRule("color-background-range")};
    ${rangeSliderCss}
  }

  &::-ms-fill-upper {
    background-color: ${themeTokenRule("color-background-range")};
    ${rangeSliderCss}
  }

  /**
   * Thumb
   */
  &::-moz-range-thumb {
    ${rangeThumbCss}
  }

  &::-ms-thumb {
    /** MS Edge fails horribly when we try to use css vars on range psuedo elements. Avoid using any CSS vars on these!  */
    border-radius: 50%;
    box-shadow: ${StaticTokenMap["color-opac-b-4"]} 0 0 0 1px inset;
    width: ${thumbWidth};
    height: ${thumbHeight};
    margin-top: ${thumbMargin};
    border: ${({ $overlay, $error }) =>
      $error ? "1px solid red" : $overlay ? "1px solid white" : "none"};
    background: ${({ $overlay }) =>
      $overlay ? "white" : StaticTokenMap["color-twitch-purple-9"]};
    appearance: none;
  }

  &::-webkit-slider-thumb {
    ${rangeThumbCss}
  }

  ${focusVisible`
    outline: none;

    &::-moz-range-thumb {
      ${rangeThumbFocusCss}
    }

    &::-ms-thumb {
      ${rangeThumbFocusCss}
    }

    &::-webkit-slider-thumb {
      ${rangeThumbFocusCss}
    }
  `}
`;

const ScRangeFill = styled.div`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  bottom: 0;
  display: flex;
  flex-direction: column;
  justify-content: center;
  z-index: ${staticTokenRule("z-index-below")};
  border-radius: ${staticTokenRule("border-radius-large")};

  ${ScRangeInput}:disabled + && {
    opacity: 0.5;
  }
`;

const ScRangeFillContainer = styled.div<{ $overlay: boolean }>`
  border-radius: ${staticTokenRule("border-radius-large")};
  background-color: ${({ $overlay }) =>
    $overlay
      ? themeTokenRule("color-background-range-overlay")
      : themeTokenRule("color-background-range")};
  width: 100%;
  height: ${trackHeight};
  cursor: pointer;
`;

const ScRangeFillValue = styled.div<{ $overlay: boolean; $error?: boolean }>`
  border-radius: ${staticTokenRule("border-radius-large")};
  background-color: ${({ $overlay, $error }) =>
    $error && $overlay
      ? themeTokenRule("color-border-overlay-error")
      : $error
      ? themeTokenRule("color-border-input-error")
      : $overlay
      ? themeTokenRule("color-background-range-overlay-fill")
      : themeTokenRule("color-background-range-fill")};
  height: 100%;
`;

export const RangeComponent: ForwardRefRenderFunction<
  HTMLInputElement,
  RangeProps
> = (props, ref) => {
  const input = (
    <ScRangeInput
      $overlay={!!props.overlay}
      $error={!!props.error}
      $filled={isRangeFilled(props)}
      {...getAriaProps(props)}
      ref={ref || props.refDelegate}
      className={"tw-range"}
      id={props.id}
      onBlur={props.onBlur}
      onChange={props.onChange}
      onFocus={props.onFocus}
      tabIndex={props.tabIndex}
      defaultValue={isRangeFilled(props) ? undefined : props.defaultValue}
      disabled={props.disabled}
      type="range"
      min={props.min}
      max={props.max}
      name={props.name}
      step={props.step}
      value={props.value}
      {...getDataProps(props)}
    />
  );

  if (isRangeFilled(props)) {
    // Calculate the percentage filled based on the min, max, and current value
    const fillPercentage =
      ((parseFloat(props.value) - props.min) / (props.max - props.min)) * 100;
    return (
      <Layout
        position={Position.Relative}
        zIndex={ZIndex.Above}
        fullWidth
        display={Display.Flex} // Necessary to cause the input to render as display: block (otherwise it has the wrong height)
      >
        {input}
        <ScRangeFill>
          <ScRangeFillContainer $overlay={!!props.overlay}>
            <ScRangeFillValue
              $overlay={!!props.overlay}
              $error={props.error}
              style={{ width: `${fillPercentage.toFixed(2)}%` }}
              data-test-selector="tw-range__fill-value-selector"
            />
          </ScRangeFillContainer>
        </ScRangeFill>
      </Layout>
    );
  }

  return input;
};

RangeComponent.displayName = "Range";
const ComponentWithRef = forwardRef(RangeComponent);
const ComponentWithOverlayContext = withOverlayContext<
  RangeProps,
  HTMLInputElement
>(ComponentWithRef);
export { ComponentWithOverlayContext as Range };
