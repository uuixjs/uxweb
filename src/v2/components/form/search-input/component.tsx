import { Display, InjectLayout } from "../../layout";
import {
  FormEventHandler,
  ForwardRefRenderFunction,
  KeyboardEventHandler,
  Ref,
  forwardRef,
} from "react";
import { Input, InputType } from "../input";
import {
  cn,
  getAriaProps,
  getDataProps,
  newUUIDv4,
} from "lib/ui-utils";

import { InputSize } from "../form";
import { SVGAsset } from "../../svg";

export interface SearchInputProps {
  autoComplete?: string;
  autoCapitalize?: boolean;
  autoFocus?: boolean;
  disabled?: boolean;
  id?: string;
  error?: boolean;
  maxLength?: number;
  name?: string;
  onBlur?: FormEventHandler<HTMLInputElement>;
  onChange?: FormEventHandler<HTMLInputElement>;
  onFocus?: FormEventHandler<HTMLInputElement>;
  /** Fires on all keypresses, including those that don't change the input value (e.g. the ↑ and ↓ keys). */
  onKeyDown?: KeyboardEventHandler<HTMLInputElement>;
  placeholder?: string;
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refDelegate` */
  refDelegate?: Ref<HTMLInputElement>;
  size?: InputSize;
  spellCheck?: boolean;
  tabIndex?: number;
  value?: string;
  /**
   * @deprecated Use `<OverlayRegion>` instead.
   */
  overlay?: boolean;
}

const SearchInput: ForwardRefRenderFunction<
  HTMLInputElement,
  SearchInputProps
> = (props, ref) => {
  const generatedId = `tw-${newUUIDv4()}`;

  return (
    <div className={cn("tw-search-input")} {...getDataProps(props)}>
      <InjectLayout display={Display.HideAccessible}>
        <label htmlFor={props.id ? props.id : generatedId}>Search</label>
      </InjectLayout>
      <Input
        {...getAriaProps(props)}
        ref={ref || props.refDelegate}
        autoCapitalize={props.autoCapitalize}
        autoFocus={props.autoFocus}
        autoComplete={props.autoComplete || "off"}
        disabled={props.disabled}
        error={props.error}
        icon={SVGAsset.NavSearch}
        id={props.id ? props.id : generatedId}
        maxLength={props.maxLength}
        name={props.name}
        onBlur={props.onBlur}
        onChange={props.onChange}
        onFocus={props.onFocus}
        onKeyDown={props.onKeyDown}
        placeholder={props.placeholder}
        spellCheck={props.spellCheck}
        tabIndex={props.tabIndex}
        size={props.size}
        type={InputType.Search}
        value={props.value}
        overlay={props.overlay}
      />
    </div>
  );
};

SearchInput.displayName = "SearchInput";
const ComponentWithRef = forwardRef(SearchInput);
export { ComponentWithRef as SearchInput };
