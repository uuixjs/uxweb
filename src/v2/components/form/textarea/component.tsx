import {
  BorderRadius,
  cn,
  getAriaProps,
  getDataProps,
  rem,
  staticTokenRule,
  styled,
} from "lib/ui-utils";
import {
  CSSProperties,
  FormEventHandler,
  ForwardRefRenderFunction,
  KeyboardEventHandler,
  Ref,
  forwardRef,
} from "react";
import { INPUT_SIZES, InputSize } from "../form";
import { InjectLayout, Overflow } from "../../layout";

import { ScInputBase } from "../common";
import { withOverlayContext } from "../../overlay-region/context";

interface ScTextAreaProps {
  $noResize?: boolean;
}

const ScTextArea = styled(ScInputBase)<ScTextAreaProps>`
  display: block;
  width: 100%;
  padding: ${staticTokenRule("space-05")} ${staticTokenRule("space-1")};
  resize: ${({ $noResize }) => ($noResize ? "none" : "vertical")};

  -ms-overflow-style: none;

  &[cols] {
    width: auto;
    max-width: 100%;
  }
`;

export interface TextAreaProps {
  autoComplete?: string;
  autoFocus?: boolean;
  cols?: number;
  disabled?: boolean;
  error?: boolean;
  id?: string;
  maxLength?: number;
  name?: string;
  onBlur?: FormEventHandler<HTMLTextAreaElement>;
  onChange?: FormEventHandler<HTMLTextAreaElement>;
  onFocus?: FormEventHandler<HTMLTextAreaElement>;
  onKeyDown?: KeyboardEventHandler<HTMLTextAreaElement>;
  overflow?: Overflow;
  paddingLeft?: number;
  paddingRight?: number;
  placeholder?: string;
  noResize?: boolean;
  readOnly?: boolean;
  rows?: number;
  size?: InputSize;
  tabIndex?: number;
  value?: string;
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refDelegate` */
  refDelegate?: Ref<HTMLTextAreaElement>;
  defaultValue?: string;
  overlay?: boolean;
}

export const TextAreaComponent: ForwardRefRenderFunction<
  HTMLTextAreaElement,
  TextAreaProps
> = (props, ref) => {
  const classesTextarea: ClassValue = {
    "tw-textarea": true,
    "tw-textarea--no-resize": props.noResize,
    "tw-textarea--error": props.error,
    "tw-textarea--overlay": props.overlay,
  };

  let inputSize = InputSize.Default;
  if (props.size) {
    inputSize = props.size;
  }

  const padding: CSSProperties = {
    paddingRight: props.paddingRight ? rem(props.paddingRight) : undefined,
    paddingLeft: props.paddingLeft ? rem(props.paddingLeft) : undefined,
  };

  return (
    <InjectLayout
      borderRadius={BorderRadius.Medium}
      overflow={props.overflow}
      fontSize={INPUT_SIZES[inputSize]}
    >
      <ScTextArea
        as="textarea"
        {...getDataProps(props)}
        {...getAriaProps(props)}
        ref={ref || props.refDelegate}
        className={cn(classesTextarea)}
        autoComplete={props.autoComplete}
        autoFocus={props.autoFocus}
        cols={props.cols}
        disabled={props.disabled}
        id={props.id}
        maxLength={props.maxLength}
        name={props.name}
        onBlur={props.onBlur}
        onChange={props.onChange}
        onFocus={props.onFocus}
        onKeyDown={props.onKeyDown}
        style={padding}
        placeholder={props.placeholder}
        rows={props.rows}
        tabIndex={props.tabIndex}
        value={props.value}
        defaultValue={props.defaultValue}
        readOnly={props.readOnly}
        error={props.error}
        overlay={props.overlay}
        $noResize={props.noResize}
      />
    </InjectLayout>
  );
};

TextAreaComponent.displayName = "TextArea";
const ComponentWithRef = forwardRef(TextAreaComponent);
const ComponentWithOverlayContext = withOverlayContext<
  TextAreaProps,
  HTMLTextAreaElement
>(ComponentWithRef);
export { ComponentWithOverlayContext as TextArea };
