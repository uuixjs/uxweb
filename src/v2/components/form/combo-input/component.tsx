import {
  CoreButton,
  CoreButtonRounding,
  CoreButtonSize,
  CoreButtonType,
} from "../../button/core-button";
import { Display, Layout } from "../../layout";
import { FC, ReactNode } from "react";
import { Input, InputProps, InputRounding } from "../input/component";
import { cn, getAriaProps, getDataProps, styled } from "lib/ui-utils";

import { CoreButtonIcon } from "../../button/core-button/core-button-icon";
import { CoreButtonLabel } from "../../button/core-button/core-button-label";
import { CoreInteractivePublicProps } from "../../core-interactive";
import { InputSize } from "../form";
import { SVGAsset } from "../../svg";

export interface ComboInputButtonProps extends CoreInteractivePublicProps {
  type?: CoreButtonType.Primary | CoreButtonType.Secondary;
  icon?: SVGAsset;
  children?: ReactNode;
}

export interface ComboInputProps extends InputProps {
  /**
   * @example { children: 'Copy' }
   */
  buttonProps: ComboInputButtonProps;
}

const ScInputWrapper = styled.div`
  flex-grow: 1;
  margin-right: 1px;
`;

const ScButtonIconWrapper = styled.div<{ $size?: InputSize }>`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: ${({ $size }) =>
    $size === InputSize.Large
      ? "0 0.5rem"
      : $size === InputSize.Small
      ? "0 0.3rem"
      : "0 0.4rem"};
`;

export const ComboInput: FC<ComboInputProps> = (props) => {
  const INPUT_BUTTON_SIZES = {
    [InputSize.Small]: CoreButtonSize.Small,
    [InputSize.Default]: CoreButtonSize.Default,
    [InputSize.Large]: CoreButtonSize.Large,
  };

  let iconClasses: ClassValue | undefined;

  if (props.buttonProps.icon && !props.buttonProps.children) {
    iconClasses = {
      "tw-combo-input__button-icon": true,
      "tw-combo-input__button-icon--small": props.size === InputSize.Small,
      "tw-combo-input__button-icon--large": props.size === InputSize.Large,
    };
  }

  let label: JSX.Element | undefined;

  if (props.buttonProps.children) {
    label = (
      <CoreButtonLabel
        icon={props.buttonProps.icon}
        size={props.size ? INPUT_BUTTON_SIZES[props.size] : undefined}
      >
        {props.buttonProps.children}
      </CoreButtonLabel>
    );
  } else if (props.buttonProps.icon) {
    label = (
      <ScButtonIconWrapper $size={props.size}>
        <CoreButtonIcon
          icon={props.buttonProps.icon}
          size={props.size ? INPUT_BUTTON_SIZES[props.size] : undefined}
        />
      </ScButtonIconWrapper>
    );
  }

  return (
    <Layout
      className="tw-combo-input"
      display={Display.Flex}
      fullWidth
      {...getDataProps(props)}
      {...getAriaProps(props)}
    >
      <ScInputWrapper className="tw-combo-input__input">
        <Input rounding={InputRounding.Left} {...props} />
      </ScInputWrapper>
      <CoreButton
        {...props.buttonProps}
        className={cn(iconClasses)}
        rounding={CoreButtonRounding.Right}
        size={props.size ? INPUT_BUTTON_SIZES[props.size] : undefined}
        variant={
          props.buttonProps.type
            ? props.buttonProps.type
            : CoreButtonType.Primary
        }
      >
        {label}
      </CoreButton>
    </Layout>
  );
};

ComboInput.displayName = "ComboInput";
