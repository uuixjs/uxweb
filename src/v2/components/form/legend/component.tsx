import { FC } from "react";
import { ScFormLabel } from "../label/component";
import { FormRequired } from "../required";

export interface FormLegendProps {
  label: string;
  required?: boolean;
}

export const FormLegend: FC<FormLegendProps> = ({
  label,
  required,
  ...props
}) => {
  return (
    <ScFormLabel as="legend" className="tw-form-legend" {...props}>
      {required && <FormRequired />}
      {label}
    </ScFormLabel>
  );
};

FormLegend.displayName = "FormLegend";
