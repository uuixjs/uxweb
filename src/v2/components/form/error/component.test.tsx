import { setupShallowTest } from "../../../tests/helpers";
import { FormErrorComponent, FormErrorProps } from "./component";

function getRequiredProps(): FormErrorProps {
  return {
    errorMessage: "The value you entered is invalid.",
  };
}

const setupShallow = setupShallowTest(FormErrorComponent, getRequiredProps);

describe("FormGroup", () => {
  it("renders form error component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });
});
