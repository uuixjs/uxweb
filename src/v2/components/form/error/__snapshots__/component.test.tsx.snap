// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormGroup renders form error component with required props 1`] = `
<Styled(Layout)
  display="flex"
  fontSize="font-size-7"
>
  <Layout
    alignItems="baseline"
    display="flex"
    height="1.5em"
    margin={
      Object {
        "right": 0.5,
      }
    }
    width="1.5em"
  >
    <Icon
      asset="NotificationError"
      fill={true}
    />
  </Layout>
  <p>
    The value you entered is invalid.
  </p>
</Styled(Layout)>
`;
